lib/PTML/CTokenEmitter.cpp

```diff
@@ -629,29 +629,57 @@ void ptml::CTokenEmitter::emitComment(llvm::StringRef Content,
   }
 }
 
+void ptml::CTokenEmitter::emitDirective(PreprocessorDirective Directive) {
+  auto Tag = PTML.initializeOpenTag(ptml::tags::Span);
+  Tag.emitAttribute(ptml::attributes::Token, ptml::c::tokens::Directive);
+  Tag.finalizeOpenTag();
+
+  switch (Directive) {
+  case PreprocessorDirective::Include:
+    return PTML.emitLiteralContent("#include");
+  case PreprocessorDirective::Pragma:
+    return PTML.emitLiteralContent("#pragma");
+  default:
+    revng_abort("Unknown preprocessor directive.");
+  }
+}
+
 void ptml::CTokenEmitter::emitIncludeDirective(llvm::StringRef Content,
                                                llvm::StringRef Location,
                                                IncludeMode Mode) {
-  // Emit include directive token:
+  emitDirective(PreprocessorDirective::Include);
+  PTML.emitLiteralContent(" ");
+
+  // Emit include path:
   {
     auto Tag = PTML.initializeOpenTag(ptml::tags::Span);
-    Tag.emitAttribute(ptml::attributes::Token, ptml::c::tokens::Directive);
+    Tag.emitAttribute(ptml::attributes::Token, ptml::c::tokens::StringLiteral);
     Tag.finalizeOpenTag();
 
-    PTML.emitLiteralContent("#include");
+    PTML.emitContent(Mode == IncludeMode::Quote ? "\"" : "<");
+    PTML.emitContent(Content);
+    PTML.emitContent(Mode == IncludeMode::Quote ? "\"" : ">");
   }
 
+  PTML.emitContentNewline();
+}
+
+void ptml::CTokenEmitter::emitPragmaDirective(llvm::StringRef Content) {
+  emitDirective(PreprocessorDirective::Pragma);
   PTML.emitLiteralContent(" ");
 
-  // Emit include path:
+  // Emit its value:
   {
     auto Tag = PTML.initializeOpenTag(ptml::tags::Span);
-    Tag.emitAttribute(ptml::attributes::Token, ptml::c::tokens::StringLiteral);
+    Tag.emitAttribute(ptml::attributes::Token, ptml::c::tokens::Constant);
     Tag.finalizeOpenTag();
 
-    PTML.emitContent(Mode == IncludeMode::Quote ? "\"" : "<");
+    revng_assert(!Content.contains('\n'));
+    for (const auto &Word : std::views::split(Content, ' '))
+      revng_assert(validateIdentifier(llvm::StringRef{ Word.begin(),
+                                                       Word.size() }));
+
```

Comment:
This is overkill for printing `#pragma once`. Hardcode it for now, and drop `emitPragmaDirective` for just `emitPragmaOnceDirective` alone. I don't see this growing in the future.